File: MCP服务器/content-extraction/AmplifyCo_botbrowser_fbdc244_typescript/python/botbrowser/converter.py
"""HTML to Markdown/text conversion."""

from __future__ import annotations

import re

from markdownify import markdownify


def html_to_markdown(html: str) -> str:
    """Convert HTML to clean markdown."""
    # Pre-process: remove images with data URIs (bloated base64) but keep normal images
    html = re.sub(r'<img[^>]+src="data:[^"]*"[^>]*/?>', "", html)

    markdown = markdownify(
        html,
        heading_style="ATX",
        bullets="-",
        convert=["a", "p", "h1", "h2", "h3", "h4", "h5", "h6",
                 "ul", "ol", "li", "table", "thead", "tbody", "tr", "th", "td",
                 "blockquote", "pre", "code", "em", "strong", "br", "hr", "img"],
    )

    # Clean up excessive whitespace
    markdown = re.sub(r"\n{3,}", "\n\n", markdown)  # Max 2 consecutive newlines
    markdown = re.sub(r"[ \t]+$", "", markdown, flags=re.MULTILINE)  # Trailing whitespace
    markdown = markdown.strip()

    return markdown


def html_to_text(html: str) -> str:
    """Convert HTML to plain text."""
    markdown = html_to_markdown(html)

    text = markdown
    text = re.sub(r"#{1,6}\s+", "", text)  # Remove heading markers
    text = re.sub(r"\*\*(.+?)\*\*", r"\1", text)  # Remove bold
    text = re.sub(r"\*(.+?)\*", r"\1", text)  # Remove italic
    text = re.sub(r"\[(.+?)\]\(.+?\)", r"\1", text)  # Remove links, keep text
    text = re.sub(r"!\[.*?\]\(.+?\)", "", text)  # Remove images
    text = re.sub(r"`{1,3}[^`]*`{1,3}", lambda m: m.group().strip("`"), text)  # Remove code markers
    text = re.sub(r"^[-*+]\s+", "", text, flags=re.MULTILINE)  # Remove list markers
    text = re.sub(r"^\d+\.\s+", "", text, flags=re.MULTILINE)  # Remove numbered list markers
    text = re.sub(r"^>\s+", "", text, flags=re.MULTILINE)  # Remove blockquote markers
    text = text.replace("---", "")  # Remove horizontal rules
    text = re.sub(r"\n{3,}", "\n\n", text)

    return text.strip()

File: reference-mcp-servers/content-extraction/AmplifyCo_botbrowser_fbdc244_typescript/js/packages/core/src/converter.ts
import TurndownService from "turndown";
import { gfm } from "turndown-plugin-gfm";

let turndownInstance: TurndownService | null = null;

function getTurndown(): TurndownService {
  if (turndownInstance) return turndownInstance;

  const turndown = new TurndownService({
    headingStyle: "atx",
    hr: "---",
    bulletListMarker: "-",
    codeBlockStyle: "fenced",
    emDelimiter: "*",
    strongDelimiter: "**",
  });

  // Add GitHub-flavored markdown support (tables, strikethrough, task lists)
  turndown.use(gfm);

  // Handle images: strip data URIs, simplify normal images
  turndown.addRule("handleImages", {
    filter: "img",
    replacement: (_content, node) => {
      const src = (node as HTMLElement).getAttribute("src") || "";
      const alt = (node as HTMLElement).getAttribute("alt") || "";
      if (!src || src.startsWith("data:")) return "";
      return `![${alt}](${src})`;
    },
  });

  turndownInstance = turndown;
  return turndown;
}

export function htmlToMarkdown(html: string): string {
  const turndown = getTurndown();
  let markdown = turndown.turndown(html);

  // Clean up excessive whitespace
  markdown = markdown
    .replace(/\n{3,}/g, "\n\n") // Max 2 consecutive newlines
    .replace(/[ \t]+$/gm, "") // Trailing whitespace
    .replace(/^\s+/, "") // Leading whitespace
    .replace(/\s+$/, ""); // Trailing whitespace

  return markdown;
}

export function htmlToText(html: string): string {
  // Use turndown but strip all markdown formatting
  const markdown = htmlToMarkdown(html);

  return markdown
    .replace(/#{1,6}\s+/g, "") // Remove heading markers
    .replace(/\*\*(.+?)\*\*/g, "$1") // Remove bold
    .replace(/\*(.+?)\*/g, "$1") // Remove italic
    .replace(/\[(.+?)\]\(.+?\)/g, "$1") // Remove links, keep text
    .replace(/!\[.*?\]\(.+?\)/g, "") // Remove images
    .replace(/`{1,3}[^`]*`{1,3}/g, (m) => m.replace(/`/g, "")) // Remove code markers
    .replace(/^[-*+]\s+/gm, "") // Remove list markers
    .replace(/^\d+\.\s+/gm, "") // Remove numbered list markers
    .replace(/^>\s+/gm, "") // Remove blockquote markers
    .replace(/---/g, "") // Remove horizontal rules
    .replace(/\n{3,}/g, "\n\n")
    .trim();
}
